Add "Intergrate tauri into your existing project" guide (#834)
tauri-apps/tauri-docs: docs/guides/getting-started/setup/integrate.mdx

---
description: If you already have an existing web project, this guide will walk you through integrating Tauri into your project
pagination_next: guides/development/development-cycle
sidebar_position: 3
sidebar_custom_props:
  doc_card_image: /img/guides/getting-started/setup/html5-light.svg
  doc_card_image_dark: /img/guides/getting-started/setup/html5-dark.svg
---

import Commands from './_fragments/_commands.mdx'
import TauriInit from './_fragments/_tauri-init.mdx'
import { InstallTauriCli, InstallTauriApi } from '@theme/Command'

# Integrate into Existing Project

If you already have an existing web project, this guide will walk you through integrating Tauri into your project, whether it is Node.js-based (like [Svelte], [React], [Vue], or [Angular]) or Rust-based (like [Yew] or [Dominator]).

:::info

Before we continue, make sure you have completed the [prerequisites] to have a working development environment.

:::

> Although Tauri is compatiable with nearly any frontend framework, we'll use examples with a [React] project throughout this guide. We'll assuming you're starting off with a project structure similar to this:

Author: ```suggestion
Although Tauri is compatiable with nearly any frontend framework, we'll use a [React] project throughout this guide created using [create-react-app]. We'll be assuming you're starting off with a project structure similar to this:
```

```
.
│── package.json
│── public
│   ╰── index.html
╰── src
    │── App.css

>     │── App.js

Author: I think react uses .jsx 
```suggestion
    │── App.jsx
```

Reviewer: You can use either with React, adding `x` allows you to use XML-style tags inside your JS from what I understand. But will take the x here

    │── index.css
    ╰── index.js
```

## Create the Rust Project

<TauriInit

>   destDirExplination='Use the path where your compiled web assets will be located. This may be something like "../dist" or "../build".'
>   devPathExplination='This is the path or address your frontend is served at during development. If servering from localhost this could be "https://localhost:3000". This can also be a directory similar to the above question.'

Author: this needs to be more decisive, no maybes, no ors.
```suggestion
  destDirExplination='use `../build` which is the path where the compiled web assets will be located.'
  devPathExplination='use `http://localhost:300` which is the url for the frontend dev server'
```

Reviewer: I tried to make this one a bit less straightforward to force the reader to use a bit more critical thinking. I feel like if we just say "use xyz" then when it doesn't work outside of a React project people won't stop to think why. @FabianLars any thoughts on this? I think we've had a few issues where people just didn't really stop and think about these values and put in an issue because of it.

Reviewer 2: Yeah, basically the same topic as here https://github.com/tauri-apps/tauri-docs/pull/834#discussion_r963604929
It must be really really obvious that these values are specific to the CRA example and may differ depending on the frontend tech.
So maybe something that explains it and/or lists common alternatives and also makes it clear that **in this case** it's ../build and localhost:3000

This was an issue wayyy too many times to not make it overly specific now :/

Reviewer: Updated the phrasing here a bit, lmk what you two think:
<img width="916" alt="Screenshot 2022-09-09 at 12 05 06" src="https://user-images.githubusercontent.com/15347255/189336470-d5920dfc-c293-4dc8-8eb4-7a3a72ebc434.png">

Reviewer 2: i like it! 👍 (there's a missing `using` at the and of the devPath explanation - "if you're **using** a different framework"

the last question here is if we need to do something similar to beforeXCommand x)

Reviewer: Just fixed that and added some explanation text for the beforexcommands

  beforeDevCommandExplination='Use "npm run dev" for this value (be sure to adapt to use your package manager of choice).'
  beforeBuildCommandExplination='Use "npm run build" for this value (be sure to adapt to use your package manager of choice).'
/>

And that's it, you have now added Tauri to your existing project and you should see a `src-tauri` directory that looks something like this:

```diff {9-14}
│── package.json
│── public
│   ╰── index.html
│── src
│   │── App.css

> │   │── App.js

Author: ditto
```suggestion
│   │── App.jsx
```

│   │── index.css
│   ╰── index.js
╰── src-tauri
    │── Cargo.toml
    │── build.rs
    │── icons
    │── src
    ╰── tauri.conf.json
```

> ## Add JavaScript API Package (Optional)

Author: we have optional here but later we require it for the commands section. we need to make two flavors of commands, one using the globalTauri and one with the npm package

Reviewer: How's this now?

Author: I was thinking more of a toggle that changes the code in the invoke example to the global tauri style.

sidebar_custom_props:
  doc_card_image: /img/guides/getting-started/setup/integrate-light.svg
  doc_card_image_dark: /img/guides/getting-started/setup/integrate-dark.svg
---

import Commands from './_fragments/_commands.mdx'
import TauriInit from './_fragments/_tauri-init.mdx'
import Command, { InstallTauriCli, InstallTauriApi } from '@theme/Command'

# Integrate into Existing Project

If you already have an existing web project, this guide will walk you through integrating Tauri into your project, whether it is Node.js-based (like [Svelte], [React], [Vue], or [Angular]) or Rust-based (like [Yew] or [Dominator]).

:::info

Before we continue, make sure you have completed the [prerequisites] to have a working development environment.

:::

Although Tauri is compatiable with nearly any frontend framework, we'll use a [React] project throughout this guide created using [create-react-app]. We'll be assuming you're starting off with a project structure similar to this:

```
.
│── package.json
│── public
│   ╰── index.html
╰── src
    │── App.css
    │── App.jsx
    │── index.css
    ╰── index.js
```

## Create the Rust Project

<TauriInit
  destDirExplination={{
    __html:

>       "If using our React example this is <code>../build</code>. Note that it may be something different like <code>../dist</code> if you're using a different framework.",

Author: ```suggestion
      "For the project example in this guide , this is <code>../build</code>. Note that it may be something different like <code>../dist</code> if you're using a different framework.",
```

  }}
  devPathExplination={{
    __html:

>       "If using our React example this is <code>http://localhost:3000</code>. Note that it may be something different (or even a directory) if you're using a different framework.",

Author: ditto

  }}
  beforeDevCommandExplination={{
    __html:

>       'If using our React example this is <code>npm run dev</code> (be sure to adapt to use your package manager of choice).',

Author: ditto

  }}
  beforeBuildCommandExplination={{
    __html:

>       'If using our React example this is <code>npm run build</code> (be sure to adapt to use your package manager of choice).',

Author: ditto

import logo from './logo.svg';
import './App.css';

import { invoke } from '@tauri-apps/api'

function App() {
  // now we can call our Command!
  // Right-click the application background and open the developer tools.
  // You will see "Hello, World!" printed in the console!
  invoke('greet', { name: 'World' })
  // `invoke` returns a Promise
  .then((response) => console.log(response))

  return (
    // -- snip --
  )
}
```

### Using `withGlobalTauri`

To interact with Tauri from your frontend without using the `@tauri-apps/api` JavaScript package you will need to enable [`withGlobalTauri`] in your `tauri.conf.json` file:

```json title=tauri.conf.json
{
  "build": {
    "beforeBuildCommand": "npm run build",
    "beforeDevCommand": "npm run dev",
    "devPath": "http://localhost:3000",
    "distDir": "../build",
    // highlight-next-line
    "withGlobalTauri": true
  },
```

This will inject a pre-bundled version of the API functions into your frontend.

You can now modify the `App.jsx` file to call your command:

```jsx title=src/App.js {4,7-12}
import logo from './logo.svg';
import './App.css';

// access the pre-bundled global API functions

> const invoke = window.__TAURI__.invoke

Author: ```suggestion
const { invoke } = window.__TAURI__
```

Reviewer: This is how it is in the HTML guide, should I adjust it over there too?

Author: probably yeah, since this is how I have it in CTA
